Calculator improvements: new operations, display formats, constants and bases

playerofgames opened the pull request:
Further to my calculator PR #5918,  I've finished my work to bring the calculator on par with other desktop calculators, adding formatting options, some missing operations, constants, and number bases. Unit tests have been updated accordingly.

### New operations
- factorial operator `!`;
- modulo operator `mod`.

### Constants added
- `PI`
- `E`

### New display format controls
- `:fix <places>` display format;
- `:sci [places]` display format;
- `:norm [display-precision]` display format;
- `:hex`, `:oct`, `:bin` and`:dec` display formats.

### New number formats
- Enter hexadecimal, octal and binary numbers using `0x`, `0o` and `0b` prefixes respectively.

### Bugs fixed
- Loss of precision when supplying numbers in scientific notation.

### About number formatting

Numbers may have practically unbounded precision, but a there is a limit of 20 decimal places in certain operations (division, roots). (Trigonometric, logarithmic and exponential operations calculate at a lower precision.)

`:norm` is now the default display format, with a _display precision_ of 21 significant digits. This tames the display of very large numbers (they will now be shown in scientific notation); the `:norm` format allows specification of greater display precision for more digits, but of course the interface is not well-suited to showing all of those digits, and we're not trying to copy Mathematica or the like.

### Demo
(using custom styling)
<img width="821" alt="Screenshot 2022-07-15 at 18 53 00" src="https://user-images.githubusercontent.com/1397817/179285513-32bf5cb2-45a1-49e1-8984-cafc076d7769.png">

I have not changed the styling of calculator in this PR.

logseq-cldwalker commented on src/main/frontend/extensions/calc.cljc:
```diff
@@ -29,28 +32,40 @@
+                                 (bn/BigNumber 'NaN')))
     :abs        (fn abs [a] (.abs a))
-    :sqrt       (fn abs [a] (.sqrt a))
+    :sqrt       (fn sqrt [a] (.sqrt a))
```
Thanks for fixing old fn names

logseq-cldwalker reviewed (changes requested):
@playerofgames Excellent work, especially with docs and testing! Some feedback on how two of the features work:

* I was initially confused by what `:norm` and `:frac` did until I read your docs. I think it would improve usability to be more explicit with these configuration commands. How about prefixing these with `:set-format` or `:format` e.g. `:set-format frac`?

* I don't know if the proposed mixed number format is that well known. I was confused by the format as underscores can have other meanings in other environments. For example, javascript just treats them as a visual delimiter and ignores them e.g. `2_2/7 -> 3.142857142857143`. Could we use a format for mixed numbers that would compute correctly in most other environments e.g. `2+2/7` and for the negative case `-2-2/7`? If that's too difficult to implement than just having a space between the integer and the proper fraction would be an improvement in readability and confusion with other environments

playerofgames commented:
Thanks for reviewing this PR.

> * ... How about prefixing these with `:set-format` or `:format` e.g. `:set-format frac`?

Agreed. Format changes are now prefixed with `:format` (or `:fmt`); I left `:hex` etc. as they are.

> * I don't know if the proposed mixed number format is that well known. I was confused by the format as underscores can have other meanings in other environments. For example, javascript just treats them as a visual delimiter and ignores them e.g. `2_2/7 -> 3.142857142857143`. Could we use a format for mixed numbers that would compute correctly in most other environments e.g. `2+2/7` and for the negative case `-2-2/7`? If that's too difficult to implement than just having a space between the integer and the proper fraction would be an improvement in readability and confusion with other environments

I agree that the mixed number format is non-standard. The alternatives that compute correctly in other environments are just expressions rather than mixed fractions. They are already available, though ungainly in the negative number case. 

The space instead of `_` is arguably better, so I've implemented this for now. It comes down to whether or not to support a mixed number _input_ format as well as an _output_ format. I think it is important that any result could be copied and incorporated into a future calculation, which is why I added the input format (and why I've 'fixed' the base output to include prefixes in this commit).

I've updated the demos and added some longer mixed number calculations. Let me know what you think!
